src/core/namespace.c

```diff
@@ -287,10 +288,18 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                // mount() always follow symlinks, so we follow symlinks during the stat() as well
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return r;
+                if (!S_ISDIR(target.st_mode))
+                        what = "/dev/null";
+                else {
+                        what = "/run/systemd/inaccessible";
+                }
                 break;
```

Comment:
well, I am pretty sure it doesn't hurt having one correct node for each node type, as it is simply the most "correct" solution to do this properly. I mean, with this proposed change in effect stat() on the masked files will generally return the same data as before, except that access rights are removed and "non-empty" nodes appear "empty" now. However, the file type stays the same. (Of course major/minor would change too, but given that they are mostly dynamically assigned on today's kernels anyway client code can't rely on fixed values for them anyway). 

Also, the option is called "InaccessiblePaths=" after all, and not "ChangeNodeTypesToDirsOrRegularPaths=" ;-). I think removing access rights and making things appear empty is pretty expected behaviour by something called "InaccessiblePaths=", however making file nodes change their node type not really...

Also, I see no real problem with maintaining one node for each type in /run/sysemd/inaccessible/... If we maintain 2 nodes or 8 doesn't really matter...